File: ipc-parser/src/main/java/com/satori/deser/domain/ConnectResponse.java
package com.satori.deser.domain;

import com.fasterxml.jackson.databind.JsonNode;
import com.satori.deser.DomainSpecificResponse;
import com.satori.deser.DomainSpecificResponseType;
import com.satori.deser.Ticket;

import java.util.Iterator;
import java.util.Optional;

//    Success(u64, u64, String),
//    Failure(u64, u64, String)
public class ConnectResponse implements DomainSpecificResponse {
    public boolean success;
    private Ticket ticket;
    public long implicated_cid;
    private String message;

    protected ConnectResponse() {}

    @Override
    public DomainSpecificResponseType getType() {
        return DomainSpecificResponseType.Connect;
    }

    @Override
    public Optional<Ticket> getTicket() {
        return Optional.of(this.ticket);
    }

    @Override
    public Optional<String> getMessage() {
        return Optional.of(this.message);
    }

    @Override
    public Optional<DomainSpecificResponse> deserializeFrom(JsonNode infoNode) {
        // debug purposes. Iterator<JsonNode> values1 = infoNode.elements();
        JsonNode successNode = infoNode.get("Success");
        if (successNode != null) {
            return deserialize_inner(successNode, true);
        }

        JsonNode failureNode = infoNode.get("Failure");
        if (failureNode != null) {
            return deserialize_inner(failureNode, false);
        }


        return Optional.empty();
    }

    private Optional<DomainSpecificResponse> deserialize_inner(JsonNode node, boolean success) {
        Iterator<JsonNode> values = node.elements();
        if (!values.hasNext()) {
            return Optional.empty();
        }

        Optional<Ticket> ticket_opt = Ticket.tryFrom(values.next().longValue());
        if (!ticket_opt.isPresent()) {
            return Optional.empty();
        }

        Ticket ticket = ticket_opt.get();

        if (!values.hasNext()) {
            return Optional.empty();
        }

        long implicated_cid = values.next().longValue();

        if (!values.hasNext()) {
            return Optional.empty();
        }

        this.success = success;
        this.ticket = ticket;
        this.message = values.next().textValue();
        this.implicated_cid = implicated_cid;
        return Optional.of(this);
    }

    public long getCID() {
        return this.implicated_cid;
    }
}

File: flutterust/packages/satori_ffi_parser/lib/types/dsr/connect_response.dart
import 'package:optional/optional.dart';
import 'package:satori_ffi_parser/types/domain_specific_response.dart';
import 'package:satori_ffi_parser/types/domain_specific_response_type.dart';
import 'package:satori_ffi_parser/types/standard_ticket.dart';

import '../u64.dart';

class RtdbConfig {
  final String apiKey;
  final String url;

  RtdbConfig(this.apiKey, this.url);
}

class ConnectResponse extends DomainSpecificResponse {
  final String message;
  final StandardTicket ticket;
  final u64 implicatedCid;
  final bool success;
  final Optional<String> jwt;
  final Optional<RtdbConfig> rtdbConfig;

  Optional<String> username = Optional.empty();

  ConnectResponse._(this.success, this.ticket, this.implicatedCid, this.message, this.jwt, this.rtdbConfig);

  @override
  Optional<String> getMessage() {
    return Optional.of(this.message);
  }

  @override
  Optional<StandardTicket> getTicket() {
    return Optional.of(this.ticket);
  }

  @override
  DomainSpecificResponseType getType() {
    return DomainSpecificResponseType.Connect;
  }

  static Optional<DomainSpecificResponse> tryFrom(Map<String, dynamic> infoNode) {
    bool success = infoNode.containsKey("Success");
    List<dynamic> leaf = infoNode[success ? "Success" : "Failure"];
    if ((success && leaf.length != 5) || (!success && leaf.length != 3)) {
      return Optional.empty();
    }

    var ticket = StandardTicket.tryFrom(leaf[0]);
    var implicatedCid = u64.tryFrom(leaf[1]);
    String message = leaf[2];

    Optional<String> jwt = success ? Optional.of(leaf[3]) : Optional.empty();
    Optional<RtdbConfig> rtdb = success ? ConnectResponse.tryParseRtdbConfig(leaf) : Optional.empty();

    return ticket.isPresent && implicatedCid.isPresent ? Optional.of(ConnectResponse._(success, ticket.value, implicatedCid.value, message, jwt, rtdb)) : Optional.empty();
  }

  static Optional<RtdbConfig> tryParseRtdbConfig(List<dynamic> leaf) {
    try {
      Map<String, dynamic> json = leaf[4];
      String apiKey = json["api_key"];
      String url = json["url"];

      return Optional.of(RtdbConfig(apiKey, url));
    } catch (_) {
      return Optional.empty();
    }
  }

  void attachUsername(String username) {
    this.username = Optional.of(username);
  }

  Optional<String> getAttachedUsername() {
    return this.username;
  }

  @override
  bool isFcm() {
    return false;
  }
}